core/src/test/java/org/elasticsearch/index/engine/KeepUntilGlobalCheckpointDeletionPolicyTests.java

```diff
@@ -0,0 +1,266 @@
+                            (long) snapshot.totalOperations(), greaterThanOrEqualTo(i + 1 - Math.max(0, globalCheckpoint.get())));
+                    }
+                }
+                engine.flush(true, true);
+            }
+            assertThat("Reserved commits should be 1", reservedCommits(globalCheckpoint.get()), hasSize(1));
+
+            try (InternalEngine engine = newEngine(store, indexPath, globalCheckpoint::get, maxSeqNoLeap::get)) {
+                assertThat("Reserved commits should always be 1", reservedCommits(globalCheckpoint.get()), hasSize(1));
+                int moreDocs = scaledRandomIntBetween(1, 100);
+                for (int i = 0; i < moreDocs; i++) {
+                    addDoc(engine, Integer.toString(initDocs + i));
+                    maxSeqNoLeap.set(randomInt(10));
+                    if (frequently()) {
+                        globalCheckpoint.set(engine.seqNoService().getLocalCheckpoint());
+                    }
+                    if (frequently()) {
+                        engine.flush(true, true);
+                        assertThat("Reserved commits should be 1", reservedCommits(globalCheckpoint.get()), hasSize(1));
+                    }
+                    if (rarely()) {
+                        engine.rollTranslogGeneration();
+                    }
+                    try (Translog.Snapshot snapshot = engine.getTranslog().newSnapshot()) {
+                        long requiredOps = initDocs + i + 1 - Math.max(0, globalCheckpoint.get());
+                        assertThat("Should keep translog operations up to the global checkpoint",
+                            (long) snapshot.totalOperations(), greaterThanOrEqualTo(requiredOps));
+                    }
+                }
+            }
+        }
+    }
+
+    public void testLegacyIndex() throws Exception {
+        final AtomicLong globalCheckpoint = new AtomicLong(randomInt(10));
+        KeepUntilGlobalCheckpointDeletionPolicy deletionPolicy = new KeepUntilGlobalCheckpointDeletionPolicy(globalCheckpoint::get);
+
+        final IndexCommit commit1 = mockIndexCommit(Collections.emptyMap());
+        deletionPolicy.onInit(Collections.singletonList(commit1));
+        verify(commit1, times(0)).delete();
+
+        final IndexCommit commit2 = mockIndexCommit(Collections.emptyMap());
+        deletionPolicy.onCommit(Arrays.asList(commit1, commit2));
+        verify(commit1, times(1)).delete();
+        verify(commit2, times(0)).delete();
+        reset(commit1, commit2);
+
+        final IndexCommit commit3 = mockIndexCommit(
+            Collections.singletonMap(SequenceNumbers.MAX_SEQ_NO, Long.toString(between(1, 1000))));
+        deletionPolicy.onCommit(Arrays.asList(commit2, commit3));
+        verify(commit2, times(0)).delete();
+        verify(commit3, times(0)).delete();
+
+        deletionPolicy.onCommit(Arrays.asList(commit1, commit2, commit3));
```

Comment:
nit - add a check that we have a good commit with a sequence numbers we keep that and drop the old ones?

Reply:
I've updated the test.